### PR TITLE
use `RefLock` in `GatedSpans` of Session

### PR DESCRIPTION
**Pull request by SparrowLii**
part of #101566 

`RefLock` is a thread-safe data structure because it can only be accessed within the thread in which it was created.

The `GatedSpans` in the Session will only be accessed in a single thread, so we can reduce the access cost in a multi-threaded environment by using `RefCell` instead of `Lock`.

### CODE REVIEW

File: compiler/rustc_data_structures/src/sync.rs [hunk 1]
```diff
@@ -460,6 +463,91 @@ impl<T: Clone> Clone for Lock<T> {
     }
 }
 
+fn next() -> NonZeroUsize {
+    static COUNTER: std::sync::atomic::AtomicUsize = std::sync::atomic::AtomicUsize::new(1);
+    NonZeroUsize::new(COUNTER.fetch_add(1, Ordering::SeqCst)).expect("more than usize::MAX threads")
```
Comment (Nilstrieb) [inline compiler/rustc_data_structures/src/sync.rs:468]:
```suggestion
    NonZeroUsize::new(COUNTER.fetch_add(1, Ordering::Relaxed)).expect("more than usize::MAX threads")
```

File: compiler/rustc_data_structures/src/sync.rs [hunk 2]
```diff
@@ -460,6 +463,91 @@ impl<T: Clone> Clone for Lock<T> {
...
+}
+
+pub(crate) fn get_thread_id() -> NonZeroUsize {
+    thread_local!(static THREAD_ID: NonZeroUsize = next());
+    THREAD_ID.with(|&x| x)
+}
+
+// `RefLock` is a thread-safe data structure because it can
+// only be used within the thread in which it was created.
```
Comment (cjgillot) [inline compiler/rustc_data_structures/src/sync.rs:477]: This comment probably deserves to be on the `unsafe impl Sync`.

File: compiler/rustc_data_structures/src/sync.rs [hunk 3]
```diff
@@ -460,6 +463,91 @@ impl<T: Clone> Clone for Lock<T> {
...
+pub struct RefLock<T> {
```
Comment (cjgillot) [inline compiler/rustc_data_structures/src/sync.rs:478]: The name is not very descriptive.
`SingleThreadAccess`?

File: compiler/rustc_data_structures/src/sync.rs [hunk 4]
```diff
@@ -460,6 +463,91 @@ impl<T: Clone> Clone for Lock<T> {
...
 
+fn next() -> NonZeroUsize {
+    static COUNTER: std::sync::atomic::AtomicUsize = std::sync::atomic::AtomicUsize::new(1);
+    NonZeroUsize::new(COUNTER.fetch_add(1, Ordering::SeqCst)).expect("more than usize::MAX threads")
+}
+
+pub(crate) fn get_thread_id() -> NonZeroUsize {
+    thread_local!(static THREAD_ID: NonZeroUsize = next());
+    THREAD_ID.with(|&x| x)
+}
```
Comment (cjgillot) [inline compiler/rustc_data_structures/src/sync.rs:474]: Don't we have something in `std::thread` or in `rayon` for this?

Comment (SparrowLii) [reply compiler/rustc_data_structures/src/sync.rs:474]: Yea, it looks like we can use `std::thread::ThreadId` here

File: compiler/rustc_data_structures/src/sync.rs [hunk 5]
```diff
@@ -460,6 +463,91 @@ impl<T: Clone> Clone for Lock<T> {
...
+
+impl<T> RefLock<T> {
+    #[inline(always)]
+    pub fn new(value: T) -> Self {
+        Self { val: RefCell::new(value), thread_id: get_thread_id() }
+    }
+
+    #[inline(always)]
+    fn assert_thread(&self) {
+        assert_eq!(get_thread_id(), self.thread_id);
```
Comment (Zoxc) [inline compiler/rustc_data_structures/src/sync.rs:491]: This check, the drop check and the `Sync` impl could be disabled when the parallel compiler is also disabled.

File: compiler/rustc_session/src/parse.rs [hunk 6]
```diff
@@ -32,7 +32,7 @@ pub type CrateCheckConfig = CheckCfg<Symbol>;
 /// used and should be feature gated accordingly in `check_crate`.
 #[derive(Default)]
 pub struct GatedSpans {
-    pub spans: Lock<FxHashMap<Symbol, Vec<Span>>>,
+    pub spans: RefLock<FxHashMap<Symbol, Vec<Span>>>,
```
Comment (Zoxc) [inline compiler/rustc_session/src/parse.rs:35]: Would this be a probably for say, parsing files in parallel?

**Comment by Zoxc** [comment]
This seems to overlap with my `OneThread` type, basically being `OneThread<RefCell<T>>`. That's missing the check on drop though.

**Comment by SparrowLii** [comment]
It looks like this runtime check still comes with a large performance penalty. I'll try some other things and in the meantime I'll turn this PR into a draft. 

We can still test the performance again by turning off the runtime checks as @Zoxc said.

**Comment by Dylan-DPC** [comment]
@SparrowLii any updates on this? thanks

**Comment by SparrowLii** [comment]
Yea I think this is an out-of-date strategy now. `RefLock` is almost same with `OneThread`, and we prefer to use `Lock` which is more efficient in #111713. As we solved the issue about performance under multi-threads, I think we would address this soon.

So I think this draft can be closed now. Thanks for your reminding!

**Pull request closed without merging**